Assets/_Scripts/Character/PlayerStats.cs

```diff
@@ -5,263 +5,407 @@
-			case (Stat.MAX_MANA):
-				return maxMana;
-			case (Stat.PHYSICAL):
-				return GetPhysical();
-			case (Stat.SPECIAL):
-				return GetSpecial();
-			case (Stat.DEFENSE):
-				return GetDefense();
-			case (Stat.RESISTANCE):
-				return GetResistance();
-			case (Stat.SPEED):
-				return GetSpeed();
-		}
-		Debug.LogWarning("Tried to retrieve invalid stat.");
-		return 0;
-	}
+    public override int GetValue() {
 
-	public void ModifyStat(Stat stat, int value) {
-		switch(stat) {
-			case (Stat.HP):
-				SetHp(GetHp() + value);
-				break;
-			case (Stat.MAX_HP):
-				maxHp += value;
-				break;
-			case (Stat.MANA):
-				SetMana(GetMana() + value);
-				break;
-			case (Stat.MAX_MANA):
-				maxMana += value;
-				break;
-			case (Stat.PHYSICAL):
-				SetPhysical(GetPhysical() + value);
-				break;
-			case (Stat.SPECIAL):
-				SetSpecial(GetSpecial() + value);
-				break;
-			case (Stat.DEFENSE):
-				SetDefense(GetDefense() + value);
-				break;
-			case (Stat.RESISTANCE):
-				SetResistance(GetResistance() + value);
+        int runningValue = baseValue;
+
+        foreach (StatModifier modifier in this.modifiers) {
+            runningValue = modifier.Apply(runningValue, this.baseValue, this.minValue, this.maxValue);
+        }
+
+        this.currentValue = runningValue;
+
+        if (this.callback != null) {
+            this.callback(this.currentValue);
+        }
+
+        return this.currentValue;
+    }
+
+    // Note: This FORCE sets the BASE value. Use sparingly.
+    public override void SetValue(int value) {
+        this.ClearModifiers();
+        this.baseValue = Mathf.Clamp(value, minValue, maxValue);
+    }
+
+    public override void ApplyDelta(int delta) {
+        this.ApplyDeltaModifier(delta, StatModifier.Type.FLAT);
+    }
+
+    public override object Clone()
+    {
+        PlayerStatWithModifiers stat = new PlayerStatWithModifiers(this.stat, this.baseValue, this.minValue, this.maxValue, this.callback);
+        PriorityList<StatModifier> clonedModifiers = new PriorityList<StatModifier>();
+
+        foreach (StatModifier modifier in this.modifiers) {
+            clonedModifiers.Add((StatModifier)modifier.Clone());
+        }
+
+        stat.modifiers = clonedModifiers;
+
+        return stat;
+    }
+
+    // Helper for simply applying change
+    public StatModifier ApplyDeltaModifier(float delta, StatModifier.Type type = StatModifier.Type.FLAT) {
+        StatModifier modifier = new StatModifier(type, delta);
+        this.ApplyModifier(modifier);
+        this.GetValue();
+        return modifier;
+    }
+
+    public int GetBaseValue() {
+        return this.baseValue;
+    }
+
+    public void SetBaseValue(int value, bool setOriginalValue = false) {
+        this.baseValue = value;
+        if (setOriginalValue) {
+            this.originalBaseValue = this.baseValue;
+        }
+    }
+
+    // Modifier helpers
+    public PriorityList<StatModifier> GetModifiers() {
+        return this.modifiers;
```

Comment:
Good practice to use `ReadOnlyCollection`

Reply:
I think PriorityList is fine here